src/main/java/co/cask/gcp/bigquery/BigQueryExecute.java

```diff
@@ -0,0 +1,230 @@
+    QueryJobConfiguration.Builder builder = QueryJobConfiguration.newBuilder(config.getSql());
+    // Run at batch priority, which won't count toward concurrent rate limit.
+    if (config.getMode().equalsIgnoreCase(MODE_BATCH)) {
+      builder.setPriority(QueryJobConfiguration.Priority.BATCH);
+    } else {
+      builder.setPriority(QueryJobConfiguration.Priority.INTERACTIVE);
+    }
+
+    // Save the results of the query to a permanent table.
+    if (config.getDataset() != null  && config.getTable() != null) {
+      builder.setDestinationTable(TableId.of(config.getDataset(), config.getTable()));
+      context.getArguments().set(context.getStageName() + ".dataset", config.getDataset());
+      context.getArguments().set(context.getStageName() + ".table", config.getTable());
+    }
+
+    // Enable or Disable the query cache to force live query evaluation.
+    if (config.shouldUseCache()) {
+      builder.setUseQueryCache(true);
+    }
+
+    // Enable legacy SQL
+    builder.setUseLegacySql(config.isLegacySQL());
+
+    QueryJobConfiguration queryConfig = builder.build();
+
+    // Location must match that of the dataset(s) referenced in the query.
+    JobId jobId = JobId.newBuilder().setRandomJob().setLocation(config.getLocation()).build();
+
+    // API request - starts the query.
+    BigQuery bigQuery = BigQueryUtils.getBigQuery(config.getServiceAccountFilePath(), config.getProject());
+    Job queryJob = bigQuery.create(
+      JobInfo.newBuilder(queryConfig)
+        .setJobId(jobId).build()
+    );
+
+    LOG.info("Executing query '%s'. The Google BigQuery job id is '%s'.", config.getSql(), jobId.getJob());
+
+    // Wait for the query to complete
+    RetryOption retryOption = RetryOption.totalTimeout(Duration.ofMinutes(config.getQueryTimeoutInMins()));
+    queryJob.waitFor(retryOption);
+
+    // Check for errors
+    if (queryJob.getStatus().getError() != null) {
+      // You can also look at queryJob.getStatus().getExecutionErrors() for all
+      // errors, not just the latest one.
+      throw new RuntimeException(queryJob.getStatus().getExecutionErrors().toString());
+    }
+
+    TableResult queryResults = queryJob.getQueryResults();
+    long rows = queryResults.getTotalRows();
+
+    context.getMetrics().gauge(RECORDS_OUT, rows);
+    context.getArguments().set(context.getStageName().concat(".query"), config.getSql());
+    context.getArguments().set(context.getStageName().concat(".jobid"), jobId.getJob());
+    context.getArguments().set(context.getStageName().concat(RECORDS_OUT), String.valueOf(rows));
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer configurer) throws IllegalArgumentException {
+    super.configurePipeline(configurer);
+
+    config.validate();
+  }
+
+  /**
+   * Config for the plugin.
+   */
+  public final class Config extends GCPConfig {
+    @Description("Use Legacy SQL.")
+    @Macro
+    private String legacy;
+
+    @Description("SQL query to execute.")
```

Comment:
descriptions and log messages are inconsistent about their use 'sql' or 'sql query'. Please figure out which is correct and use it everywhere.